pkg/ruler/compat.go

```diff
@@ -143,12 +143,17 @@ func EngineQueryFunc(engine *promql.Engine, q storage.Queryable, overrides Rules
 	}
 }
 
-func MetricsQueryFunc(qf rules.QueryFunc, queries, failedQueries prometheus.Counter) rules.QueryFunc {
+func MetricsQueryFunc(qf rules.QueryFunc, queries, failedQueries prometheus.Counter, queryTime prometheus.Counter) rules.QueryFunc {
 	return func(ctx context.Context, qs string, t time.Time) (promql.Vector, error) {
 		queries.Inc()
 
-		result, err := qf(ctx, qs, t)
+		// If we've been passed a counter we want to record the wall time spent executing this request.
+		if queryTime != nil {
```

Comment:
This should be moved out to a separate `QueryFunc`. That way we don't break public signature of `MetricsQueryFunc`, and we only need to apply this function if query stats is enabled. (See how `MetricsQueryFunc` and `EngineQueryFunc` are composed below).

(I know `MetricsQueryFunc` is tempting, as it's just another metric, but the purpose of that perhaps wrongly-named function was to track global query metrics and failures).